File: apps/web/src/features/match-replay/model/weaponTier.ts
/**
 * weaponTier.ts — À QUEL NIVEAU SE LIT UNE ARME DE SOCLE : base, terrain, puissance, non classé.
 *
 * CE QUE LE BLOC NE SAVAIT PAS DIRE. « Contrôle des armes spéciales » comptait toutes les prises
 * de socle à égalité : reprendre son fusil d'assaut posé sur un râtelier y pesait autant que
 * rafler le lance-roquettes du socle central. Ce n'est pas ce qu'un match raconte. Demande
 * utilisateur du 2026-09-13 : « faire la distinction entre les armes de base, les armes de
 * terrain et les armes spéciales […] sur les râteliers ce sont des armes de terrain, sur les
 * socles des armes de puissance ».
 *
 * LES TROIS SOURCES, ET AUCUNE N'EST UN NOM D'ARME :
 *
 *   BASE       l'arme est dans l'équipement de DÉPART d'une vie du match — canal `loadouts`,
 *              PREMIÈRE émission de chaque `slot` (un slot = une vie).
 *   TERRAIN    l'emplacement de la carte qui confirme le socle est un RÂTELIER (`rack`).
 *   PUISSANCE  cet emplacement est un SOCLE DE PUISSANCE (`power`).
 *   NON CLASSÉ aucun emplacement ne confirme le socle. Il reste VISIBLE avec son compte,
 *              jamais fondu dans un autre niveau.
 *
 * LA NATURE D'UN EMPLACEMENT VIENT DE LA CARTE, JAMAIS DE L'ARME. Mesure du 2026-09-14 sur 76
 * artefacts : 70 socles sur 669 portent une arme de rôle « lourd » (Hydra, Needler, Sentinel
 * Beam, Shock Rifle) sur un RÂTELIER, et c'est nominal. Juger au nom de l'arme produirait donc
 * 10 % de faux niveaux ; la même Hydra est de terrain sur une carte et de puissance sur une
 * autre.
 *
 * JUMEAU DE `internal/analysis/weapontier` (Go), QUI SERT LES AGRÉGATS. Les deux existent parce
 * que la résolution vit de deux côtés : ici la vue match, qui a la catégorie de mode dans son
 * en-tête ; là-bas les pages Sessions / Escouade / Timeseries, qui agrègent côté serveur. Les
 * deux partagent les mêmes seuils écrits et les mêmes témoins — toute divergence est un bug,
 * pas une variante.
 *
 * Pur : aucun React, aucune couleur, AUCUNE LANGUE. Les intertitres vivent dans l'i18n du rejeu.
 */
import type { ReplayDocumentReady } from '@/lib/replay/replayNormalize'

/**
 * Les niveaux, dans l'ORDRE DE LECTURE du bloc.
 *
 * PUISSANCE EN TÊTE DEPUIS LE 2026-09-21 (décision utilisateur, amendant D2) : puissance,
 * puis terrain, puis base dans un dépliable fermé. Ce qui décide un match ouvre le bloc ;
 * reprendre son fusil d'assaut ferme la marche.
 *
 * `powerup` ET `unclassified` FERMENT LA LISTE ET NE SE RENDENT PLUS (même décision, cf.
 * `MatchPadControlSection`) : un socle de BONUS est un ÉQUIPEMENT, il se lit dans « Usages
 * d'équipement » (`EPISODE_FAMILIES`, `equipmentUsageLogic.ts`) et n'a rien à faire dans le
 * contrôle des ARMES ; « non identifié » est une absence de mesure, pas un niveau de jeu. Le
 * CLASSEMENT (`padTierOf`) est inchangé — seul l'affichage retire ces deux groupes.
 *
 * IL NE DÉPARTAGE PLUS RIEN : voir `PAD_TIER_TIEBREAK` ci-dessous.
 */
export const PAD_TIER_ORDER = ['power', 'ground', 'base', 'powerup', 'unclassified'] as const

export type PadTier = (typeof PAD_TIER_ORDER)[number]

/**
 * ORDRE DE DÉPARTAGE d'une arme vue à deux niveaux dans le même match (`tierOfWeaponOf`) —
 * DISTINCT de l'ordre d'affichage, et c'est tout l'objet de cette constante.
 *
 * DÉCOUPLAGE DU 2026-09-21 (lot G), sur découverte du lot D. Les deux usages lisaient
 * `PAD_TIER_ORDER` : inverser l'ordre de LECTURE du bloc (base → puissance en tête, commit
 * a9733a985) a donc silencieusement inversé le DÉPARTAGE d'un conflit à compte égal, sans
 * qu'aucune décision produit ne le demande. Un ordre d'écran est un choix de mise en page ;
 * un ordre de départage est une règle de classement — ils n'ont pas à bouger ensemble.
 *
 * La valeur ci-dessous RESTAURE le départage d'avant a9733a985 (`base, ground, power,
 * powerup, unclassified`). Le conflit est rare et mesuré : UNE arme sur 59 matchs et 2 881
 * prises (0,17 %), `terrain` contre `non classé` (cf. `padControlLogic.ts`).
 */
export const PAD_TIER_TIEBREAK: readonly PadTier[] = [
  'base',
  'ground',
  'power',
  'powerup',
  'unclassified',
]

/**
 * Part minimale des vies du match qu'une arme doit occuper AU DÉPART pour être tenue pour une
 * arme de base.
 *
 * POURQUOI UN SEUIL, ET POURQUOI CELUI-LÀ (mesure du 2026-09-14, 4 406 vies en classé et partie
 * rapide). Les trois armes de départ réelles pèsent 94,4 % des équipements de première émission
 * (MA40 45,5 %, Sidekick 29,5 %, BR75 19,4 %) ; une queue de 5,6 % porte des armes qui n'en sont
 * pas — Skewer, Sniper, Épée — parce que la première émission publiée d'une vie arrive parfois
 * après un ramassage. Sans seuil, un Skewer vu une fois au départ d'une vie ferait passer TOUTES
 * ses prises en « base » et viderait le niveau « puissance » de son sens. À 5 % la queue tombe
 * entièrement, et les trois vraies armes de base passent avec quatre à neuf fois la marge.
 *
 * La valeur est celle de `weapontier.BaseShareMin` côté Go. Elle ne retire jamais une prise du
 * bloc : une arme recalée retombe sur le niveau de son emplacement.
 */
export const BASE_SHARE_MIN = 0.05

/** Ce qu'un match apprend une fois pour toutes sur ses niveaux. */
export interface PadTierMatch {
  /** Famille d'emplacement par index de socle (`weaponPads`) ; `null` = non confirmé. */
  familyByPad: (string | null)[]
  /** Familles d'arme retenues comme armes de départ (hexadécimal, clé de `weaponPads[].weapon`). */
  baseWeapons: Set<string>
  /** Le mode distribue des départs aléatoires : pas de niveau « base ». */
  randomStarts: boolean
  /** Nombre de vies dont l'équipement de départ a été lu (dénominateur de `BASE_SHARE_MIN`). */
  lives: number
  /**
   * FAUX quand AUCUN emplacement de la carte n'a confirmé de socle — carte hors référence, ou
   * repère décalé. Le bloc doit alors dire « niveaux non établis », et surtout PAS ranger tout
   * le match sous « Non classé » : l'un est une absence de mesure, l'autre un résultat de
   * mesure, et les confondre ferait lire un défaut de carte comme un fait de jeu.
   */
  tiersMeasured: boolean
}

/**
 * Assemble le classement d'un match.
 *
 * LE CARACTÈRE ALÉATOIRE DES DÉPARTS VIENT DU SERVEUR (`doc.weaponTiers.randomStarts`), jamais
 * d'une liste de catégories côté web. Cette liste a existé une semaine et a divergé : la règle
 * vit dans le TOML du titre, elle gouverne aussi l'écriture en base, et deux copies auraient
 * fini par ranger le même match dans deux niveaux différents selon la page (revue 2026-09-14).
 *
 * Champ absent = le serveur ne sait pas (mode inconnu, titre sans règle) : on ne publie alors
 * PAS la note « départs aléatoires », et on ne l'invente pas non plus — le niveau « base » se
 * mesure, et il se vérifie de lui-même.
 */
export function buildPadTierMatch(doc: ReplayDocumentReady): PadTierMatch {
  const familyByPad: (string | null)[] = doc.weaponPads.map(() => null)
  for (const spot of doc.mapWeaponPads?.pads ?? []) {
    if (spot.pad >= 0 && spot.pad < familyByPad.length) familyByPad[spot.pad] = spot.family
  }
  const randomStarts = doc.weaponTiers?.randomStarts === true
  const { weapons, lives } = baseWeaponsOf(doc)
  return {
    familyByPad,
    baseWeapons: randomStarts ? new Set<string>() : weapons,
    randomStarts,
    lives,
    tiersMeasured: familyByPad.some((f) => f !== null),
  }
}

/**
 * Le niveau d'UNE prise : le socle où elle a eu lieu, et l'arme qui s'y trouvait.
 *
 * L'ORDRE DES TESTS EST LA RÈGLE : base > terrain > puissance > non classé. Une arme de départ
 * ramassée sur un râtelier se lit « base » — reprendre son fusil d'assaut n'est pas contrôler
 * la carte. Un socle de BONUS échappe à cet ordre : ce n'est pas une arme.
 */
export function padTierOf(match: PadTierMatch, padIndex: number, weapon: string): PadTier {
  const family = match.familyByPad[padIndex] ?? null
  if (family === 'powerup') return 'powerup'
  if (!match.randomStarts && match.baseWeapons.has(weapon)) return 'base'
  if (family === 'rack') return 'ground'
  if (family === 'power') return 'power'
  return 'unclassified'
}

/**
 * baseWeaponsOf — les armes de départ du match.
 *
 * UNE SEULE ÉMISSION COMPTE PAR SLOT, LA PREMIÈRE, et c'est mesuré : un `slot` est une VIE (100
 * slots distincts pour 113 pistes sur le match témoin `01e1f945`), et le canal RÉ-ÉMET en cours
 * de vie après un changement d'arme. Prendre tout le canal dilue les trois armes de base de
 * 94,4 % à 85,6 % en classé et fait monter le S7 Sniper de 0,73 % à 3,15 % — autrement dit, il
 * transforme une arme de puissance en arme de base.
 */
function baseWeaponsOf(doc: ReplayDocumentReady): { weapons: Set<string>; lives: number } {
  const premiere = new Map<number, string[]>()
  for (const lo of doc.loadouts) {
    if (!premiere.has(lo.slot)) premiere.set(lo.slot, lo.w)
  }
  const lives = premiere.size
  const weapons = new Set<string>()
  if (lives === 0) return { weapons, lives }
  const compte = new Map<string, number>()
  for (const w of premiere.values()) {
    // Une même arme deux fois dans le même équipement de départ ne vaut qu'UNE vie : le
    // dénominateur est la vie, pas l'emplacement d'inventaire.
    for (const arme of new Set(w)) {
      if (!arme) continue
      compte.set(arme, (compte.get(arme) ?? 0) + 1)
    }
  }
  const seuil = lives * BASE_SHARE_MIN
  for (const [arme, n] of compte) {
    if (n >= seuil) weapons.add(arme)
  }
  return { weapons, lives }
}

File: apps/go-api/internal/analysis/weapontier/weapontier.go
// Package weapontier décide À QUEL NIVEAU se lit une arme prise sur un socle : arme de
// BASE, arme de TERRAIN, arme de PUISSANCE, ou non classée.
//
// POURQUOI CE PAQUET EXISTE. Le bloc « contrôle des armes » comptait jusqu'ici des prises de
// socle toutes égales entre elles : ramasser un fusil d'assaut posé sur un râtelier et
// ramasser le lance-roquettes du socle central y pesaient pareil. Ce n'est pas ce qu'un match
// raconte. Le niveau remet chaque prise à sa place, et il se MESURE — il ne s'écrit pas dans
// une liste d'armes.
//
// LES TROIS SOURCES, ET AUCUNE N'EST UN NOM D'ARME :
//
//	BASE       l'arme est dans l'équipement de DÉPART d'une vie du match (canal `loadouts`
//	           du film, PREMIÈRE émission de chaque slot, et SEULEMENT si aucune prise
//	           d'arme de cette vie ne la précède — cf. [Take]). Donc BR75/Bandit en classé,
//	           AR+Sidekick en partie rapide, sans qu'aucune liste ne le déclare.
//	TERRAIN    l'emplacement de la carte qui confirme le socle est un RÂTELIER (`rack`).
//	PUISSANCE  cet emplacement est un SOCLE DE PUISSANCE (`power`).
//	NON CLASSÉ aucun emplacement ne confirme le socle (carte hors référence, ou socle hors
//	           du rayon d'un mètre). Il reste VISIBLE avec son compte — jamais fondu dans un
//	           autre niveau (décision D4 du plan).
//
// LA NATURE D'UN EMPLACEMENT VIENT DE LA CARTE, JAMAIS DE L'ARME (décision D1). Mesure de
// l'étape 0 sur 76 artefacts : 70 socles sur 669 portent une arme de rôle « lourd » (Hydra,
// Needler, Sentinel Beam, Shock Rifle) sur un RÂTELIER, et c'est nominal. Brancher le niveau
// sur le rôle du registre canonique produirait donc 10 % de faux niveaux.
//
// PUR : aucune base, aucun HTTP, aucun log, aucune langue. Le paquet compte et classe ; c'est
// l'appelant qui journalise (cf. [CrossCheck]) et l'écran qui nomme.
package weapontier

// Tier est le niveau de lecture d'une arme de socle.
type Tier string

const (
	// TierBase — arme présente dans un équipement de départ du match.
	TierBase Tier = "base"
	// TierGround — arme apparue sur un râtelier de la carte.
	TierGround Tier = "ground"
	// TierPower — arme apparue sur un socle de puissance.
	TierPower Tier = "power"
	// TierPowerup — socle de BONUS (surbouclier, camouflage). Ce n'est pas une arme, et ce
	// niveau existe pour que le bloc puisse l'écarter SANS tester un préfixe de nom : le
	// compte des bonus vit à part depuis toujours (`powerupOccupations`), et ce paquet ne
	// change pas cette frontière, il la nomme.
	TierPowerup Tier = "powerup"
	// TierUnclassified — socle qu'aucun emplacement de la carte ne confirme.
	TierUnclassified Tier = "unclassified"
)

// Familles d'emplacement telles que la référence des cartes les écrit
// (`MapWeaponPadSpot.Family`, dérivée du `type_id` Forge).
const (
	familyRack    = "rack"
	familyPower   = "power"
	familyPowerup = "powerup"
)

// BaseShareMin est la part minimale des vies du match qu'une arme doit occuper au départ pour
// être tenue pour une arme de BASE.
//
// POURQUOI UN SEUIL SUBSISTE APRÈS LE CORRECTIF DU 2026-09-21. La queue d'armes qui n'en sont
// pas venait d'une CAUSE, et cette cause est désormais traitée en amont : le canal `loadouts`
// est publié sur une GRILLE D'IMAGES-CLÉS GLOBALE (une émission toutes les 200 frames = 20 s,
// témoin `b1ad85eb` : t = 12, 212, 412, … 5014), si bien que la première émission d'une vie
// accuse 0 à 19,2 s de retard sur l'apparition — le temps d'un ramassage. [baseWeaponsOf]
// écarte maintenant toute émission POSTÉRIEURE à une prise d'arme de la même vie (cf. [Take]).
//
// Le seuil reste parce que les canaux de prise ne couvrent pas TOUTES les prises (le témoin
// b1ad85eb publie 118 prises d'arme pour 87 vies, et deux vies y prennent un Empaleur sans
// qu'aucun canal ne l'écrive — seule la chaîne des objets au sol le trahit). Mesure du
// 2026-09-21 sur le parc des 92 artefacts rangés : après le correctif, la queue plafonne à
// 4,94 % des vies retenues et la plus faible vraie arme de base tient 5,06 % — 5 % sépare
// encore les deux populations, sans marge à revoir tant que cet écart tient. À re-mesurer si
// la couverture des canaux de prise change.
//
// Ce seuil ne s'applique QU'À la promotion en « base » : il ne retire jamais une prise du
// tableau, elle retombe simplement sur le niveau de son emplacement.
const BaseShareMin = 0.05

// Pad est UN socle du match, reduit a ce que le niveau demande : la famille d'arme qui s'y
// trouve. Sa POSITION dans la tranche est son identite — c'est l'index que [Spot.Pad] cite et
// que [Match.TierOf] recoit.
type Pad struct {
	Weapon string
}

// Spot est UN emplacement de la carte CONFIRME par un socle du match : la nature que le fichier
// de carte lui donne, et l'index du socle qu'elle qualifie. Un socle qu'aucun Spot ne cite n'est
// pas confirme — son niveau sera « non classe ».
type Spot struct {
	Pad int
	// Family : `rack`, `power` ou `powerup`, tels que la reference des cartes les ecrit.
	Family string
}

// Spawn est UNE emission d'equipement de depart : la vie concernee, l'instant de l'emission et
// les armes en main.
//
// L'ORDRE DE LA TRANCHE COMPTE : seule la PREMIERE emission de chaque Slot est retenue (cf.
// [baseWeaponsOf]). L'appelant passe le canal dans l'ordre du film, il ne le trie pas.
type Spawn struct {
	Slot uint32
	// T est l'instant de l'emission, dans l'unite de l'appelant (frames cote film). Il n'est
	// compare qu'a des [Take] de la MEME vie : ce paquet n'a pas besoin de connaitre l'unite.
	T       int
	Weapons []string
}

// Take est UNE PRISE D'ARME EN COURS DE VIE : la vie concernee et son instant.
//
// CE QU'ELLE SERT A FAIRE, ET RIEN D'AUTRE : disqualifier une emission d'equipement de depart
// qui la SUIT. Une emission posterieure a une prise ne dit plus ce que la vie portait au
// depart, elle dit ce que le joueur a ramasse — c'est exactement ce qui classait l'Empaleur en
// arme de base sur le match `b1ad85eb` (constat utilisateur du 2026-09-21).
//
// LA DOTATION DE REAPPARITION N'EN EST PAS UNE : le jeu remet les armes de depart en main a
// l'apparition, et certains canaux l'ecrivent comme une prise. C'est a l'appelant d'ecarter ce
// qui n'est pas posterieur au DEBUT de la vie — lui seul sait ou commence une vie.
type Take struct {
	Slot uint32
	T    int
}

// Match porte ce qu'un match apprend une fois pour toutes : la nature de chaque socle et les
// armes de départ. Le construire coûte une passe ; classer une prise coûte ensuite un accès.
//
// Le zéro est UTILISABLE : tous les socles y sont non classés et aucune arme n'est de base.
// C'est exactement ce qu'un match sans référence de carte et sans canal `loadouts` doit
// donner.
type Match struct {
	// familyByPad est indexé par l'index de socle de `weaponPads` ; "" = non confirmé.
	familyByPad []string
	// baseWeapons : familles d'arme (hexadécimal, même écriture que `WeaponPad.Weapon`)
	// retenues comme armes de départ du match.
	baseWeapons map[string]bool
	// randomStarts — le mode distribue des équipements de départ ALÉATOIRES (Fiesta et
	// consorts). Le niveau « base » n'y a aucun sens et n'est jamais attribué.
	randomStarts bool
	// lives est le nombre de vies dont un équipement de départ a été lu — le DÉNOMINATEUR de
	// BaseShareMin, et rien d'autre. Il n'est pas exposé : ce que l'écran affiche vient des
	// compteurs du bloc servi, pas d'un accesseur de ce paquet (0 code mort, revue 2026-09-14).
	lives int
}

// NewMatch assemble le classement d'un match.
//
// `randomStarts` est REÇU, jamais déduit : c'est la catégorie de mode qui le sait
// (Fiesta / Super Fiesta / Husky Raid), et ce paquet ne lit ni slug ni nom de mode.
func NewMatch(pads []Pad, spots []Spot, spawns []Spawn, takes []Take, randomStarts bool) Match {
	m := Match{familyByPad: make([]string, len(pads)), randomStarts: randomStarts}
	for _, spot := range spots {
		if spot.Pad >= 0 && spot.Pad < len(m.familyByPad) {
			m.familyByPad[spot.Pad] = spot.Family
		}
	}
	if !randomStarts {
		m.baseWeapons, m.lives = baseWeaponsOf(spawns, takes)
	} else {
		_, m.lives = baseWeaponsOf(spawns, takes)
	}
	return m
}

// FamilyOfPad rend la famille d'emplacement d'un socle du match ("" si non confirmé).
func (m Match) FamilyOfPad(padIndex int) string {
	if padIndex < 0 || padIndex >= len(m.familyByPad) {
		return ""
	}
	return m.familyByPad[padIndex]
}

// TierOf classe UNE prise : le socle où elle a eu lieu, et l'arme qui s'y trouvait.
//
// L'ORDRE DES TESTS EST LA RÈGLE DU PLAN : base > terrain > puissance > non classé. Une arme
// de départ ramassée sur un râtelier se lit « base » — c'est le niveau d'information voulu :
// reprendre son fusil d'assaut n'est pas contrôler la carte.
//
// Un socle de BONUS échappe à cet ordre : il n'est pas une arme, et il ne devient jamais
// « base » même si une famille de bonus portait par accident le nom d'une arme de départ.
func (m Match) TierOf(padIndex int, weapon string) Tier {
	family := m.FamilyOfPad(padIndex)
	if family == familyPowerup {
		return TierPowerup
	}
	if !m.randomStarts && m.baseWeapons[weapon] {
		return TierBase
	}
	switch family {
	case familyRack:
		return TierGround
	case familyPower:
		return TierPower
	default:
		return TierUnclassified
	}
}

// baseWeaponsOf lit les armes de départ du match dans le canal `loadouts`.
//
// UNE SEULE ÉMISSION COMPTE PAR SLOT, LA PREMIÈRE, et c'est mesuré (étape 0) : un `slot` est
// une VIE (100 slots distincts pour 113 pistes sur le match témoin `01e1f945`), et le canal
// RÉ-ÉMET en cours de vie après un changement d'arme. Prendre tout le canal dilue les trois
// armes de base de 94,4 % à 85,6 % en classé et fait monter le S7 Sniper de 0,73 % à 3,15 % —
// c'est-à-dire qu'il transforme une arme de puissance en arme de base.
//
// ET CETTE PREMIÈRE ÉMISSION N'EST PAS UNE LECTURE D'APPARITION. Le canal est publié sur une
// grille d'images-clés GLOBALE, pas au spawn : sur `b1ad85eb`, les 25 instants d'émission sont
// t = 12, 212, 412 … 5014, un toutes les 200 frames, et l'écart entre le début d'une vie et sa
// première émission va de 0 à 192 frames (19,2 s ; médiane 60). Ce qui est lu là n'est donc
// l'équipement de départ QUE SI rien n'a été ramassé entre-temps. Une vie dont la première
// émission SUIT une prise d'arme ([Take]) n'apprend rien sur les départs : elle quitte le
// numérateur ET le dénominateur — le silence n'est pas un zéro.
//
// MESURE DU CORRECTIF (2026-09-21, 92 artefacts rangés) : 6 385 vies retenues sur 7 844 ; 34
// matchs changent d'ensemble d'armes de base, et les faux départs tombent (b1ad85eb :
// l'Empaleur passe de 6,85 % à 4,25 % des vies retenues, donc sous le seuil, tandis que MA40
// et Sidekick montent de 86,3 % à 97,9 % et 93,6 %).
//
// Rend les familles retenues et le nombre de vies lues.
func baseWeaponsOf(spawns []Spawn, takes []Take) (map[string]bool, int) {
	premierePrise := make(map[uint32]int, len(takes))
	for _, t := range takes {
		if vu, ok := premierePrise[t.Slot]; !ok || t.T < vu {
			premierePrise[t.Slot] = t.T
		}
	}
	premiere := make(map[uint32]int, len(spawns))
	ecartees := make(map[uint32]bool)
	for i, l := range spawns {
		if _, vu := premiere[l.Slot]; vu || ecartees[l.Slot] {
			continue
		}
		if prise, ok := premierePrise[l.Slot]; ok && l.T >= prise {
			ecartees[l.Slot] = true
			continue
		}
		premiere[l.Slot] = i
	}
	vies := len(premiere)
	if vies == 0 {
		return nil, 0
	}
	compte := make(map[string]int)
	for _, i := range premiere {
		// Une même arme deux fois dans le même équipement de départ ne vaut qu'une vie :
		// le dénominateur est la VIE, pas l'emplacement d'inventaire.
		vues := make(map[string]bool, len(spawns[i].Weapons))
		for _, w := range spawns[i].Weapons {
			if w == "" || vues[w] {
				continue
			}
			vues[w] = true
			compte[w]++
		}
	}
	seuil := float64(vies) * BaseShareMin
	out := make(map[string]bool)
	for w, n := range compte {
		if float64(n) >= seuil {
			out[w] = true
		}
	}
	return out, vies
}
